src/SignalR/clients/csharp/Http.Connections.Client/src/Internal/WebSocketsTransport.cs

```diff
@@ -278,26 +286,90 @@ public async Task StartAsync(Uri url, TransferFormat transferFormat, Cancellatio
 
         Log.StartedTransport(_logger);
 
-        // Create the pipe pair (Application's writer is connected to Transport's reader, and vice versa)
-        var pair = DuplexPipe.CreateConnectionPair(_httpConnectionOptions.TransportPipeOptions, _httpConnectionOptions.AppPipeOptions);
+        _stopCts = new CancellationTokenSource();
 
-        _transport = pair.Transport;
-        _application = pair.Application;
+        var ignoreFirstCanceled = false;
+
+        if (_transport is null)
+        {
+            // Create the pipe pair (Application's writer is connected to Transport's reader, and vice versa)
+            DuplexPipePair pair;
+            if (_useAck)
+            {
+                pair = CreateAckConnectionPair(_httpConnectionOptions.TransportPipeOptions, _httpConnectionOptions.AppPipeOptions);
+            }
+            else
+            {
+                pair = DuplexPipe.CreateConnectionPair(_httpConnectionOptions.TransportPipeOptions, _httpConnectionOptions.AppPipeOptions);
+            }
+
+            _transport = pair.Transport;
+            _application = pair.Application;
+        }
+        else
+        {
+            if (_application!.Input is AckPipeReader reader)
+            {
+                if (reader.Resend())
+                {
+                    // Start reconnect ack handshake
+                    // 1. Send ack ID to server for last message we recieved from server before we disconnected
+                    // 2. Read from server to get the last ack ID it received before we disconnecting
+                    // 3. Resume normal send/receive loops
+
+                    ignoreFirstCanceled = true;
+                    var buf = new byte[AckPipeWriter.FrameSize];
+                    AckPipeWriter.WriteFrame(buf.AsSpan(), 0, ((AckPipeWriter)_transport.Output).LastAck);
+                    await _webSocket.SendAsync(new ArraySegment<byte>(buf, 0, AckPipeWriter.FrameSize), _webSocketMessageType, true, _stopCts.Token).ConfigureAwait(false);
+
+                    Array.Clear(buf, 0, buf.Length);
+                    var receiveResult = await _webSocket.ReceiveAsync(new ArraySegment<byte>(buf, 0, AckPipeWriter.FrameSize), _stopCts.Token).ConfigureAwait(false);
+                    // server sends 0 length, but with latest ack, so there shouldn't be more than a frame of data sent
+                    Debug.Assert(receiveResult.Count == AckPipeWriter.FrameSize);
```

Comment:
IMO this should be a hard test, not just a `Debug.Assert` - it is subject to external influence

Reply:
actually, no this definitely doesn't cover it - since we don't slice, `ParseFrame` would see 24 bytes *regardless* of how many we actually received; you might also need to to a receive loop here, i.e.

``` c#
int totalRead = 0;
while (totalRead < AckPipeWriter.FrameSize)
{
    var receiveResult = await _webSocket.ReceiveAsync(new ArraySegment<byte>(buf, totalRead, AckPipeWriter.FrameSize - totalRead), _stopCts.Token).ConfigureAwait(false);
    if (receiveResult.Count <= 0) break; // EOF
    totalRead += receiveResult.Count;
}
if (totalRead < AckPipeWriter.FrameSize) throw new EndOfStreamException(); // didn't receive header
```

*now* we know that we've read the 24-byte header